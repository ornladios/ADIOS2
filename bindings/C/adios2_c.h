/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * adios2_c.h : ADIOS2 C bindings declarations
 *
 *  Created on: Mar 13, 2017
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_BINDINGS_C_ADIOS2_C_H_
#define ADIOS2_BINDINGS_C_ADIOS2_C_H_

#include "adios2/ADIOSConfig.h"

#include "c/adios2_c_adios.h"
#include "c/adios2_c_attribute.h"
#include "c/adios2_c_engine.h"
#include "c/adios2_c_io.h"
#include "c/adios2_c_operator.h"
#include "c/adios2_c_variable.h"

#endif /* ADIOS2_BINDINGS_C_ADIOS2_C_H_ */
