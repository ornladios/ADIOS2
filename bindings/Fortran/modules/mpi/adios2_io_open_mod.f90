!
! Distributed under the OSI-approved Apache License, Version 2.0.  See
!  accompanying file Copyright.txt for details.
!
!  adios2_io_open_mod.f90 : ADIOS2 Fortran bindings for IO class open function
!
!   Created on: Mar 13, 2017
!       Author: William F Godoy godoywf@ornl.gov
!

module adios2_io_open_mod
    use adios2_parameters_mod
    implicit none

    interface adios2_open
        module procedure adios2_open_old_comm
        module procedure adios2_open_new_comm
    end interface

contains

    subroutine adios2_open_old_comm(engine, io, name, adios2_mode, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(out) :: ierr

        engine%mode = adios2_mode
        call adios2_open_f2c(engine%f2c, io%f2c, TRIM(ADJUSTL(name))//char(0), &
                             adios2_mode, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

    subroutine adios2_open_new_comm(engine, io, name, adios2_mode, comm, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(in) :: comm
        integer, intent(out) :: ierr

        call adios2_open_new_comm_f2c(engine%f2c, io%f2c, &
                                      TRIM(ADJUSTL(name))//char(0), &
                                      adios2_mode, comm, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

end module
