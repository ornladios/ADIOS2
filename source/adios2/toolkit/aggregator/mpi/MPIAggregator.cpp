/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIAggregator.cpp
 *
 *  Created on: Feb 20, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#include "MPIAggregator.h"

#include "adios2/helper/adiosFunctions.h"

namespace adios2
{
namespace aggregator
{

MPIAggregator::MPIAggregator() : m_Comm(MPI_COMM_SELF) {}

MPIAggregator::~MPIAggregator()
{
    if (m_IsActive)
    {
        helper::CheckMPIReturn(MPI_Comm_free(&m_Comm),
                               "freeing aggregators comm in MPIAggregator "
                               "destructor, not recommended");
    }
}

void MPIAggregator::Init(const size_t subStreams, MPI_Comm parentComm) {}

void MPIAggregator::SwapBuffers(const int step) noexcept {}

void MPIAggregator::ResetBuffers() noexcept {}

BufferSTL &MPIAggregator::GetConsumerBuffer(BufferSTL &bufferSTL)
{
    return bufferSTL;
}

void MPIAggregator::Close()
{
    if (m_IsActive)
    {
        helper::CheckMPIReturn(MPI_Comm_free(&m_Comm),
                               "freeing aggregators comm at Close\n");
        m_IsActive = false;
    }
}

// PROTECTED
void MPIAggregator::InitComm(const size_t subStreams, MPI_Comm parentComm)
{
    int parentRank;
    int parentSize;
    MPI_Comm_rank(parentComm, &parentRank);
    MPI_Comm_size(parentComm, &parentSize);

    const size_t processes = static_cast<size_t>(parentSize);
    size_t stride = processes / subStreams + 1;
    const size_t remainder = processes % subStreams;

    size_t consumer = 0;

    for (auto s = 0; s < subStreams; ++s)
    {
        if (s >= remainder)
        {
            stride = processes / subStreams;
        }

        if (static_cast<size_t>(parentRank) >= consumer &&
            static_cast<size_t>(parentRank) < consumer + stride)
        {
            helper::CheckMPIReturn(
                MPI_Comm_split(parentComm, static_cast<int>(consumer),
                               parentRank, &m_Comm),
                "creating aggregators comm with split at Open");
            m_ConsumerRank = static_cast<int>(consumer);
            m_SubStreamIndex = static_cast<size_t>(s);
        }

        consumer += stride;
    }

    MPI_Comm_rank(m_Comm, &m_Rank);
    MPI_Comm_size(m_Comm, &m_Size);

    if (m_Rank != 0)
    {
        m_IsConsumer = false;
    }

    m_IsActive = true;
    m_SubStreams = subStreams;
}

void MPIAggregator::HandshakeRank(const int rank)
{
    int message = -1;
    if (m_Rank == rank)
    {
        message = m_Rank;
    }

    helper::CheckMPIReturn(MPI_Bcast(&message, 1, MPI_INT, rank, m_Comm),
                           "handshake with aggregator rank 0 at Open");
}

} // end namespace aggregator
} // end namespace adios2
