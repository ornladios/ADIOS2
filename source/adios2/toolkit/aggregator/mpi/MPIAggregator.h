/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIAggregator.h
 *
 *  Created on: Feb 20, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPIAGGREGATOR_H_
#define ADIOS2_TOOLKIT_AGGREGATOR_MPI_MPIAGGREGATOR_H_

#include "adios2/ADIOSMPI.h"
#include "adios2/ADIOSTypes.h"
#include "adios2/toolkit/format/BufferSTL.h"

namespace adios2
{
namespace aggregator
{

class MPIAggregator
{
public:
    /** total number of substreams */
    size_t m_SubStreams = 0;

    /** current substream index from 0 to m_SubStreams-1 */
    size_t m_SubStreamIndex = 0;

    /** split Communicator for a substream: producers and consumer (rank=0) */
    MPI_Comm m_Comm;

    /** rank from m_Comm */
    int m_Rank = 0;

    /** size from m_Comm */
    int m_Size = 1;

    /**
     * true: consumes data from itself or other processes and interacts with
     * transport managers
     */
    bool m_IsConsumer = true;

    /** true: doing aggregation, false: not doing aggregation */
    bool m_IsActive = false;

    /** consumer original rank coming from the parent communicator in Init,
     *  corresponds to m_Rank = 0 */
    int m_ConsumerRank = -1;

    MPIAggregator();

    virtual ~MPIAggregator();

    virtual void Init(const size_t subStreams, MPI_Comm parentComm);

    virtual std::vector<std::vector<MPI_Request>>
    IExchange(BufferSTL &bufferSTL, const int step) = 0;

    virtual std::vector<std::vector<MPI_Request>>
    IExchangeAbsolutePosition(BufferSTL &bufferSTL, const int step) = 0;

    virtual void
    WaitAbsolutePosition(std::vector<std::vector<MPI_Request>> &requests,
                         const int step) = 0;

    virtual void Wait(std::vector<std::vector<MPI_Request>> &requests,
                      const int step) = 0;

    virtual void SwapBuffers(const int step) noexcept;

    virtual void ResetBuffers() noexcept;

    virtual BufferSTL &GetConsumerBuffer(BufferSTL &bufferSTL);

    /** closes current aggregator, frees m_Comm */
    void Close();

protected:
    /** Init m_Comm splitting assigning ranks to subStreams (balanced except for
     * the last rank) */
    void InitComm(const size_t subStreams, MPI_Comm parentComm);

    /** handshakes a single rank with the rest of the m_Comm ranks */
    void HandshakeRank(const int rank = 0);

    /** assigning extra buffers for aggregation */
    std::vector<BufferSTL> m_Buffers;
};

} // end namespace aggregator
} // end namespace adios2

#endif /* ADIOS2_TOOLKIT_AGGREGATOR_MPIAGGREGATOR_H_ */
