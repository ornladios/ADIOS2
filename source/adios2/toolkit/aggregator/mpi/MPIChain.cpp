/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * MPIChain.cpp
 *
 *  Created on: Feb 21, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */
#include "MPIChain.h"

#include "adios2/ADIOSMPI.h"
#include "adios2/helper/adiosFunctions.h" //helper::CheckMPIReturn

namespace adios2
{
namespace aggregator
{

MPIChain::MPIChain() : MPIAggregator() {}

void MPIChain::Init(const size_t subStreams, MPI_Comm parentComm)
{
    InitComm(subStreams, parentComm);
    HandshakeRank(0);
    HandshakeLinks();

    // add a receiving buffer except for the last rank (only sends)
    if (m_Rank < m_Size)
    {
        m_Buffers.emplace_back(); // just one for now
    }
}

std::vector<std::vector<MPI_Request>> MPIChain::IExchange(BufferSTL &bufferSTL,
                                                          const int step)
{
    if (m_Size == 1)
    {
        return std::vector<std::vector<MPI_Request>>();
    }

    BufferSTL &sendBuffer = GetSender(bufferSTL);
    const int endRank = m_Size - 1 - step;
    const bool sender = (m_Rank >= 1 && m_Rank <= endRank) ? true : false;
    const bool receiver = (m_Rank < endRank) ? true : false;

    std::vector<std::vector<MPI_Request>> requests(2);

    if (sender) // sender
    {
        requests[0].resize(1);

        helper::CheckMPIReturn(MPI_Isend(&sendBuffer.m_Position, 1,
                                         ADIOS2_MPI_SIZE_T, m_Rank - 1, 0,
                                         m_Comm, &requests[0][0]),
                               ", aggregation Isend size at iteration " +
                                   std::to_string(step) + "\n");

        // only send data if buffer larger than 0
        if (sendBuffer.m_Position > 0)
        {

            const std::vector<MPI_Request> requestsISend64 =
                helper::Isend64(sendBuffer.m_Buffer.data(),
                                sendBuffer.m_Position, m_Rank - 1, 1, m_Comm,
                                ", aggregation Isend64 data at iteration " +
                                    std::to_string(step));

            requests[0].insert(requests[0].end(), requestsISend64.begin(),
                               requestsISend64.end());
        }
    }
    // receive size, resize receiving buffer and receive data
    if (receiver)
    {
        size_t bufferSize = 0;
        MPI_Request receiveSizeRequest;
        helper::CheckMPIReturn(MPI_Irecv(&bufferSize, 1, ADIOS2_MPI_SIZE_T,
                                         m_Rank + 1, 0, m_Comm,
                                         &receiveSizeRequest),
                               ", aggregation Irecv size at iteration " +
                                   std::to_string(step) + "\n");

        MPI_Status receiveStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&receiveSizeRequest, &receiveStatus),
            ", aggregation waiting for receiver size at iteration " +
                std::to_string(step) + "\n");

        BufferSTL &receiveBuffer = GetReceiver(bufferSTL);
        ResizeUpdateBufferSTL(
            bufferSize, receiveBuffer,
            "in aggregation, when resizing receiving buffer to size " +
                std::to_string(bufferSize));

        // only receive data if buffer is larger than 0
        if (bufferSize > 0)
        {
            requests[1] =
                helper::Irecv64(receiveBuffer.m_Buffer.data(),
                                receiveBuffer.m_Position, m_Rank + 1, 1, m_Comm,
                                ", aggregation Irecv64 data at iteration " +
                                    std::to_string(step));
        }
    }

    return requests;
}

std::vector<std::vector<MPI_Request>>
MPIChain::IExchangeAbsolutePosition(BufferSTL &bufferSTL, const int step)
{
    if (m_Size == 1)
    {
        return std::vector<std::vector<MPI_Request>>();
    }

    if (m_IsInExchangeAbsolutePosition)
    {
        throw std::runtime_error("ERROR: MPIChain::IExchangeAbsolutePosition: "
                                 "An existing exchange is still active.");
    }

    const int destination = (step != m_Size - 1) ? step + 1 : 0;
    std::vector<std::vector<MPI_Request>> requests(2,
                                                   std::vector<MPI_Request>(1));

    if (step == 0)
    {
        m_SizeSend =
            (m_Rank == 0) ? bufferSTL.m_AbsolutePosition : bufferSTL.m_Position;
    }

    if (m_Rank == step)
    {
        m_ExchangeAbsolutePosition =
            (m_Rank == 0) ? m_SizeSend
                          : m_SizeSend + bufferSTL.m_AbsolutePosition;

        helper::CheckMPIReturn(
            MPI_Isend(&m_ExchangeAbsolutePosition, 1, ADIOS2_MPI_SIZE_T,
                      destination, 0, m_Comm, &requests[0][0]),
            ", aggregation Isend absolute position at iteration " +
                std::to_string(step) + "\n");
    }
    else if (m_Rank == destination)
    {
        helper::CheckMPIReturn(
            MPI_Irecv(&bufferSTL.m_AbsolutePosition, 1, ADIOS2_MPI_SIZE_T, step,
                      0, m_Comm, &requests[1][0]),
            ", aggregation Irecv absolute position at iteration " +
                std::to_string(step) + "\n");
    }

    m_IsInExchangeAbsolutePosition = true;
    return requests;
}

void MPIChain::Wait(std::vector<std::vector<MPI_Request>> &requests,
                    const int step)
{
    if (m_Size == 1)
    {
        return;
    }

    const int endRank = m_Size - 1 - step;
    const bool sender = (m_Rank >= 1 && m_Rank <= endRank) ? true : false;
    const bool receiver = (m_Rank < endRank) ? true : false;

    MPI_Status status;
    if (receiver)
    {
        for (auto &req : requests[1])
        {
            helper::CheckMPIReturn(
                MPI_Wait(&req, &status),
                ", aggregation waiting for receiver request at iteration " +
                    std::to_string(step) + "\n");
        }
    }

    if (sender)
    {
        for (auto &req : requests[0])
        {
            helper::CheckMPIReturn(
                MPI_Wait(&req, &status),
                ", aggregation waiting for sender request at iteration " +
                    std::to_string(step) + "\n");
        }
    }
}

void MPIChain::WaitAbsolutePosition(
    std::vector<std::vector<MPI_Request>> &requests, const int step)
{
    if (m_Size == 1)
    {
        return;
    }

    if (!m_IsInExchangeAbsolutePosition)
    {
        throw std::runtime_error("ERROR: MPIChain::WaitAbsolutePosition: An "
                                 "existing exchange is not active.");
    }

    MPI_Status status;
    const int destination = (step != m_Size - 1) ? step + 1 : 0;

    if (m_Rank == destination)
    {
        helper::CheckMPIReturn(
            MPI_Wait(&requests[1][0], &status),
            ", aggregation Irecv Wait absolute position at iteration " +
                std::to_string(step) + "\n");
    }

    if (m_Rank == step)
    {
        helper::CheckMPIReturn(
            MPI_Wait(&requests[0][0], &status),
            ", aggregation Isend Wait absolute position at iteration " +
                std::to_string(step) + "\n");
    }
    m_IsInExchangeAbsolutePosition = false;
}

void MPIChain::SwapBuffers(const int /*step*/) noexcept
{
    m_CurrentBufferOrder = (m_CurrentBufferOrder == 0) ? 1 : 0;
}

void MPIChain::ResetBuffers() noexcept { m_CurrentBufferOrder = 0; }

BufferSTL &MPIChain::GetConsumerBuffer(BufferSTL &bufferSTL)
{
    return GetSender(bufferSTL);
}

// PRIVATE
void MPIChain::HandshakeLinks()
{
    int link = -1;

    MPI_Request sendRequest;
    if (m_Rank > 0) // send
    {
        helper::CheckMPIReturn(
            MPI_Isend(&m_Rank, 1, MPI_INT, m_Rank - 1, 0, m_Comm, &sendRequest),
            "Isend handshake with neighbor, MPIChain aggregator, at Open");
    }

    if (m_Rank < m_Size - 1) // receive
    {
        MPI_Request receiveRequest;
        helper::CheckMPIReturn(
            MPI_Irecv(&link, 1, MPI_INT, m_Rank + 1, 0, m_Comm,
                      &receiveRequest),
            "Irecv handshake with neighbor, MPIChain aggregator, at Open");

        MPI_Status receiveStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&receiveRequest, &receiveStatus),
            "Irecv Wait handshake with neighbor, MPIChain aggregator, at Open");
    }

    if (m_Rank > 0)
    {
        MPI_Status sendStatus;
        helper::CheckMPIReturn(
            MPI_Wait(&sendRequest, &sendStatus),
            "Isend wait handshake with neighbor, MPIChain aggregator, at Open");
    }
}

BufferSTL &MPIChain::GetSender(BufferSTL &bufferSTL)
{
    if (m_CurrentBufferOrder == 0)
    {
        return bufferSTL;
    }
    else
    {
        return m_Buffers.front();
    }
}

BufferSTL &MPIChain::GetReceiver(BufferSTL &bufferSTL)
{
    if (m_CurrentBufferOrder == 0)
    {
        return m_Buffers.front();
    }
    else
    {
        return bufferSTL;
    }
}

void MPIChain::ResizeUpdateBufferSTL(const size_t newSize, BufferSTL &bufferSTL,
                                     const std::string hint)
{
    bufferSTL.Resize(newSize, hint);
    bufferSTL.m_Position = bufferSTL.m_Buffer.size();
}

} // end namespace aggregator
} // end namespace adios2
