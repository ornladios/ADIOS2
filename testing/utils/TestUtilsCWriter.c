/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * TestUtilsCWriter
 *
 *  Created on: Sept 19, 2018
 *      Author: Norbert Podhorszki
 */

#include <stdio.h>
#include <string.h>

#include <adios2_c.h>

#ifdef ADIOS2_HAVE_MPI
#include <mpi.h>
#endif

#include "SmallTestData_c.h"

int main(int argc, char *argv[])
{
    int rank = 0;
    int nproc = 1;
#ifdef ADIOS2_HAVE_MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &nproc);
    adios2_adios *adiosH = adios2_init(MPI_COMM_WORLD, adios2_debug_mode_on);
#else
    adios2_adios *adiosH = adios2_init(adios2_debug_mode_on);
#endif

    char engineName[32] = "BPFile";
    if (argc > 1)
    {
        strncpy(engineName, argv[1], sizeof(engineName));
    }

    // IO
    adios2_io *ioH = adios2_declare_io(adiosH, "CArrayTypes");
    // Set engine parameters
    adios2_set_engine(ioH, engineName);
    adios2_set_parameter(ioH, "ProfileUnits", "Microseconds");
    adios2_set_parameter(ioH, "Threads", "1");

    // 1D shape
    size_t shape[1], start[1], count[1];
    shape[0] = nproc * data_Nx;
    start[0] = rank * data_Nx;
    count[0] = data_Nx;

    // 2D shape
    size_t shape2[2], start2[2], count2[2];
    shape2[0] = nproc * d2_Nx;
    shape2[1] = d2_Ny;
    start2[0] = rank * d2_Nx;
    start2[1] = 0;
    count2[0] = d2_Nx;
    count2[1] = d2_Ny;

    // Define variables in ioH
    adios2_define_variable(ioH, "nproc", adios2_type_int32_t, 0, NULL, NULL,
                           NULL, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varI8", adios2_type_int8_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI16", adios2_type_int16_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI32", adios2_type_int32_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varI64", adios2_type_int64_t, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varU8", adios2_type_uint8_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU16", adios2_type_uint16_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU32", adios2_type_uint32_t, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varU64", adios2_type_uint64_t, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "varR32", adios2_type_float, 1, shape, start,
                           count, adios2_constant_dims_true);
    adios2_define_variable(ioH, "varR64", adios2_type_double, 1, shape, start,
                           count, adios2_constant_dims_true);

    adios2_define_variable(ioH, "R64_2d", adios2_type_double, 2, shape2, start2,
                           count2, adios2_constant_dims_true);

    // Define attributes in ioH
    adios2_define_attribute(ioH, "name", adios2_type_string, "TestUtilsCWrite");
    adios2_define_attribute_array(ioH, "strarray", adios2_type_string, strarray,
                                  sizeof(strarray) / sizeof(char *));
    adios2_define_attribute_array(ioH, "nwriters", adios2_type_int32_t, &nproc,
                                  1);
    unsigned short shape2D[2] = {(unsigned short)d2_Nx, (unsigned short)d2_Ny};
    adios2_define_attribute_array(ioH, "shape2D", adios2_type_uint16_t, shape2D,
                                  2);
    adios2_define_attribute_array(ioH, "aI8", adios2_type_int8_t, data_I8, 1);
    adios2_define_attribute_array(ioH, "aI16", adios2_type_int16_t, data_I16,
                                  1);
    adios2_define_attribute_array(ioH, "aI32", adios2_type_int32_t, data_I32,
                                  1);
    adios2_define_attribute_array(ioH, "aU8", adios2_type_uint8_t, data_U8, 1);
    adios2_define_attribute_array(ioH, "aU16", adios2_type_uint16_t, data_U16,
                                  1);
    adios2_define_attribute_array(ioH, "aU32", adios2_type_uint32_t, data_U32,
                                  1);
    adios2_define_attribute_array(ioH, "aR32", adios2_type_float, data_R32, 1);
    adios2_define_attribute_array(ioH, "aR64", adios2_type_double, data_R64, 1);

    // inquire variables
    adios2_variable *varNproc = adios2_inquire_variable(ioH, "nproc");
    adios2_variable *varI8 = adios2_inquire_variable(ioH, "varI8");
    adios2_variable *varI16 = adios2_inquire_variable(ioH, "varI16");
    adios2_variable *varI32 = adios2_inquire_variable(ioH, "varI32");
    adios2_variable *varI64 = adios2_inquire_variable(ioH, "varI64");
    adios2_variable *varU8 = adios2_inquire_variable(ioH, "varU8");
    adios2_variable *varU16 = adios2_inquire_variable(ioH, "varU16");
    adios2_variable *varU32 = adios2_inquire_variable(ioH, "varU32");
    adios2_variable *varU64 = adios2_inquire_variable(ioH, "varU64");
    adios2_variable *varR32 = adios2_inquire_variable(ioH, "varR32");
    adios2_variable *varR64 = adios2_inquire_variable(ioH, "varR64");
    adios2_variable *R64_2d = adios2_inquire_variable(ioH, "R64_2d");

    adios2_engine *engineH =
        adios2_open(ioH, "TestUtilsCWriter.bp", adios2_mode_write);

    adios2_put(engineH, varNproc, &nproc, adios2_mode_deferred);
    adios2_put(engineH, varI8, data_I8, adios2_mode_deferred);
    adios2_put(engineH, varI16, data_I16, adios2_mode_deferred);
    adios2_put(engineH, varI32, data_I32, adios2_mode_deferred);
    adios2_put(engineH, varI64, data_I64, adios2_mode_deferred);

    adios2_put(engineH, varU8, data_U8, adios2_mode_deferred);
    adios2_put(engineH, varU16, data_U16, adios2_mode_deferred);
    adios2_put(engineH, varU32, data_U32, adios2_mode_deferred);
    adios2_put(engineH, varU64, data_U64, adios2_mode_deferred);

    adios2_put(engineH, varR32, data_R32, adios2_mode_deferred);
    adios2_put(engineH, varR64, data_R64, adios2_mode_deferred);

    adios2_put(engineH, R64_2d, d2_R64, adios2_mode_deferred);
    adios2_close(engineH);

    // deallocate adiosH
    adios2_finalize(adiosH);

#ifdef ADIOS2_HAVE_MPI
    MPI_Finalize();
#endif

    return 0;
}
