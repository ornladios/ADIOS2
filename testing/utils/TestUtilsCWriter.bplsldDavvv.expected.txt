Settings :
  masks  : 0 
  file   : TestUtilsCWriter.bp
  output : stdout
      -l : show scalar values and min/max/avg of arrays
      -a : list attributes too
      -d : dump matching variables and attributes
      -D : show decomposition of variables in the file

ADIOS Open: read header info from TestUtilsCWriter.bp
Try BPFile engine to open the file...
File info:
  of variables:  12
  of attributes: 12
  statistics:    Min / Max 

  double    R64_2d    {4, 5} = 0 / 19
        step 0: 
          block 0: [0:3, 0:4] = 0 / 19
    j=0, st=0 ct=4
    s[0]=0, c[0]=4, n=4
    j=1, st=0 ct=5
    s[1]=0, c[1]=5, n=20
 total size of data to read = 160
Read size strategy:
    dim 1: read 5 elements
    dim 0: read 4 elements
    read 20 elements at once, 20 in total (nelems=20)
adios_read_var name=R64_2d   start = { 0 0 }  count = { 4 5 }  read 20 elems
    read block 0 from offset 0 nelems 20)
set selection:   start = { 0 0 }  count = { 4 5 }
    (0,0)    0 1 2 3 4 5
    (1,1)    6 7 8 9 10 11
    (2,2)    12 13 14 15 16 17
    (3,3)    18 19 
  int16_t   aI16      attr   = 512
  int32_t   aI32      attr   = 131072
  int8_t    aI8       attr   = 0
  float     aR32      attr   = 0
  double    aR64      attr   = 0
  uint16_t  aU16      attr   = 32768
  uint32_t  aU32      attr   = 2147483648
  uint8_t   aU8       attr   = 128
  string    name      attr   = "TestUtilsCWrite"
  int32_t   nproc     scalar = 1
        step 0:  = 1
               1
  int32_t   nwriters  attr   = 1
  uint16_t  shape2D   attr   = {4, 5}
  string    strarray  attr   = {"first", "second", "third", "fourth"}
  int16_t   varI16    {10} = -510 / 521
        step 0: 
          block 0: [0:9] = -510 / 521
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 20
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varI16   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    512 513 -510 515 -508 517
    (6)    -506 519 -504 521 
  int32_t   varI32    {10} = -131070 / 131081
        step 0: 
          block 0: [0:9] = -131070 / 131081
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 40
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varI32   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    131072 131073 -131070 131075 -131068 131077
    (6)    -131066 131079 -131064 131081 
  int64_t   varI64    {10} = -8589934590 / 8589934601
        step 0: 
          block 0: [0:9] = -8589934590 / 8589934601
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 80
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varI64   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    8589934592 8589934593 -8589934590 8589934595 -8589934588 8589934597
    (6)    -8589934586 8589934599 -8589934584 8589934601 
  int8_t    varI8     {10} = -8 / 9
        step 0: 
          block 0: [0:9] = -8 / 9
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 10
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varI8   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    0 1 -2 3 -4 5
    (6)    -6 7 -8 9 
  float     varR32    {10} = 0 / 9
        step 0: 
          block 0: [0:9] = 0 / 9
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 40
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varR32   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    0 1 2 3 4 5
    (6)    6 7 8 9 
  double    varR64    {10} = 0 / 9
        step 0: 
          block 0: [0:9] = 0 / 9
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 80
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varR64   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    0 1 2 3 4 5
    (6)    6 7 8 9 
  uint16_t  varU16    {10} = 32768 / 32777
        step 0: 
          block 0: [0:9] = 32768 / 32777
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 20
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varU16   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    32768 32769 32770 32771 32772 32773
    (6)    32774 32775 32776 32777 
  uint32_t  varU32    {10} = 2147483648 / 2147483657
        step 0: 
          block 0: [0:9] = 2147483648 / 2147483657
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 40
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varU32   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    2147483648 2147483649 2147483650 2147483651 2147483652 2147483653
    (6)    2147483654 2147483655 2147483656 2147483657 
  uint64_t  varU64    {10} = 9223372036854775808 / 9223372036854775817
        step 0: 
          block 0: [0:9] = 9223372036854775808 / 9223372036854775817
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 80
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varU64   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    9223372036854775808 9223372036854775809 9223372036854775810 9223372036854775811 9223372036854775812 9223372036854775813
    (6)    9223372036854775814 9223372036854775815 9223372036854775816 9223372036854775817 
  uint8_t   varU8     {10} = 128 / 137
        step 0: 
          block 0: [0:9] = 128 / 137
    j=0, st=0 ct=10
    s[0]=0, c[0]=10, n=10
 total size of data to read = 10
Read size strategy:
    dim 0: read 10 elements
    read 10 elements at once, 10 in total (nelems=10)
adios_read_var name=varU8   start = { 0 }  count = { 10 }  read 10 elems
    read block 0 from offset 0 nelems 10)
set selection:   start = { 0 }  count = { 10 }
    (0)    128 129 130 131 132 133
    (6)    134 135 136 137 
